### ISSUE TITLE
Conditional validation

### ISSUE THREAD

**Issue by bartoszlenar** [bartoszlenar/CoreValidation#28]
### Current behavior

This will record error if `NotEmpty` rule reports failure:

``` csharp
.Member(m => m.SomeMember, m => m.NotEmpty())
```

### Expected behavior

Would be nice to have conditional validation (an additional condition based on which the validation of the particular member is triggered).

Still not sure about the design, so lets have several examples:

#### Single rule

`When` (or `WithCondition` to be more consistent with the proposed `WithName` #27 and existing `WithMessage`) predicate will be checked before stepping into `NotWhiteSpace` rule and if false, the rule will not be even executed against the member.

``` csharp
.Member(m => m.SomeMember, m => 
        m.NotWhiteSpace().When(parentModel => parentModel.SomeMember != "")
)
```

❓ should the parent model be used in condition predicate? or rather the member itself?

``` csharp
.Member(m => m.SomeMember, m => 
        m.NotWhiteSpace().When(someMember=> someMember != "")
)
```

❓ should we have several options available for all cases?

* `When(member => member != "")`
* `When(m => m.AnotherMember, m => m.NotEmpty())`
* `WhenParent(parent => parent.SomeMember != "")`

#### Scopes

Attached to the scope should work exactly the same:

``` csharp
.Member(m => m.SomeMember, m => m
                .NotWhiteSpace().NotEmpty().MaxLength(30)
        ).When(model=> model != "")
```
